README.md
# cram-md5-digest-js<!-- omit in toc -->

[![Build Status](https://travis-ci.org/Constantiner/cram-md5-digest-js.svg?branch=master)](https://travis-ci.org/Constantiner/cram-md5-digest-js) [![codecov](https://codecov.io/gh/Constantiner/cram-md5-digest-js/branch/master/graph/badge.svg)](https://codecov.io/gh/Constantiner/cram-md5-digest-js)

[CRAM-MD5](https://en.wikipedia.org/wiki/CRAM-MD5) digester implementation in JavaScript with zero dependencies.

It uses a digest string from server and generates a hash using a secret (usually, password).

The library supports converting the resulting hash to Base64-encoded string.

- [Installation](#installation)
- [Changelog](#changelog)
- [Usage](#usage)
- [Documentation](#documentation)
	- [cramMd5Digest](#crammd5digest)
	- [cramMd5DigestBase64](#crammd5digestbase64)

## Installation

```
npm install @constantiner/cram-md5-digest
```

## Changelog

Versions [changelog](CHANGELOG.md).

## Usage

```javascript
import { cramMd5DigestBase64 } from "@constantiner/cram-md5-digest";

cramMd5DigestBase64("voxjx70j131lv9g645j6Znx8535l5Xb1nkl6CF0489lE2Gc^b*H&F5jn", "1$G15/iw");
```

## Documentation

### cramMd5Digest

```javascript
import { cramMd5Digest } from "@constantiner/cram-md5-digest";
```

Generates MD5 hash from `passwordString` and `cramKey`.

It has the following parameters:

* `passwordString` Is a secret (password) to hash (`String`).
* `cramKey` Is a key (digest) to hash password with (`String`).

It returns a 32 symbols length string with calculated hash.

### cramMd5DigestBase64

```javascript
import { cramMd5DigestBase64 } from "@constantiner/cram-md5-digest";
```

Generates MD5 hash from `passwordString` and `cramKey` and encodes it in Base64.

It has the following parameters:

* `passwordString` Is a secret (password) to hash (`String`).
* `cramKey` Is a key (digest) to hash password with (`String`).
* `padding` (optional, default value is `false`) if `true` there is Base64 padding (=) added.

It returns a string with Base64-encoded hash (with padding or not).


CHANGELOG.md
# Changelog

## [v0.9.8](https://github.com/Constantiner/cram-md5-digest-js/tree/v0.9.8) (2020-05-08)

[Full Changelog](https://github.com/Constantiner/cram-md5-digest-js/compare/0.9.7...v0.9.8)

This release improves all the project infrastructure to meet all the best practices. Library's sources are written in TypeScript now and build contains corresponding type definitions. We drop support for UMD build. Anyway, all the APIs haven't changed.

**Implemented enhancements:**

- Update documentation [\#28](https://github.com/Constantiner/cram-md5-digest-js/issues/28)
- Add badges to README [\#24](https://github.com/Constantiner/cram-md5-digest-js/issues/24)
- Add TypeScript support [\#15](https://github.com/Constantiner/cram-md5-digest-js/issues/15)
- Include actual current year into license text [\#14](https://github.com/Constantiner/cram-md5-digest-js/issues/14)
- Add CI support [\#13](https://github.com/Constantiner/cram-md5-digest-js/issues/13)
- Add changelog generation [\#12](https://github.com/Constantiner/cram-md5-digest-js/issues/12)
- Revise Husky and lint-staged configuration [\#11](https://github.com/Constantiner/cram-md5-digest-js/issues/11)
- Extract Prettier configuration into separate files [\#10](https://github.com/Constantiner/cram-md5-digest-js/issues/10)
- Add cool ESLint plugins to improve code quality [\#9](https://github.com/Constantiner/cram-md5-digest-js/issues/9)
- Add watch mode support for tests [\#8](https://github.com/Constantiner/cram-md5-digest-js/issues/8)
- Use Rollup for build [\#7](https://github.com/Constantiner/cram-md5-digest-js/issues/7)

**Security fixes:**

- Update dependencies [\#18](https://github.com/Constantiner/cram-md5-digest-js/issues/18)

**Merged pull requests:**

- Update documentation [\#30](https://github.com/Constantiner/cram-md5-digest-js/pull/30)
- Add cool ESLint plugins to improve code quality [\#29](https://github.com/Constantiner/cram-md5-digest-js/pull/29)
- Add TypeScript support [\#27](https://github.com/Constantiner/cram-md5-digest-js/pull/27)
- Add changelog generation [\#26](https://github.com/Constantiner/cram-md5-digest-js/pull/26)
- Add badges to README [\#25](https://github.com/Constantiner/cram-md5-digest-js/pull/25)
- Add CI support [\#23](https://github.com/Constantiner/cram-md5-digest-js/pull/23)
- Add watch mode support for tests [\#22](https://github.com/Constantiner/cram-md5-digest-js/pull/22)
- Revise Husky and lint-staged configuration [\#21](https://github.com/Constantiner/cram-md5-digest-js/pull/21)
- Extract Prettier configuration into separate files [\#20](https://github.com/Constantiner/cram-md5-digest-js/pull/20)
- Update dependencies [\#19](https://github.com/Constantiner/cram-md5-digest-js/pull/19)
- Use Rollup for build [\#17](https://github.com/Constantiner/cram-md5-digest-js/pull/17)

## [0.9.7](https://github.com/Constantiner/cram-md5-digest-js/tree/0.9.7) (2019-02-08)

[Full Changelog](https://github.com/Constantiner/cram-md5-digest-js/compare/0.9.6...0.9.7)

## [0.9.6](https://github.com/Constantiner/cram-md5-digest-js/tree/0.9.6) (2019-02-08)

[Full Changelog](https://github.com/Constantiner/cram-md5-digest-js/compare/0.9.5...0.9.6)

**Implemented enhancements:**

- Create changelog file [\#3](https://github.com/Constantiner/cram-md5-digest-js/issues/3)
- Format non minified distribution files with Prettier [\#2](https://github.com/Constantiner/cram-md5-digest-js/issues/2)
- Include license text in distribution files header [\#1](https://github.com/Constantiner/cram-md5-digest-js/issues/1)

**Merged pull requests:**

- Create changelog file [\#6](https://github.com/Constantiner/cram-md5-digest-js/pull/6)
- Format non minified distribution files with Prettier [\#5](https://github.com/Constantiner/cram-md5-digest-js/pull/5)
- Include license text in distribution files header [\#4](https://github.com/Constantiner/cram-md5-digest-js/pull/4)

## [0.9.5](https://github.com/Constantiner/cram-md5-digest-js/tree/0.9.5) (2018-10-09)

[Full Changelog](https://github.com/Constantiner/cram-md5-digest-js/compare/0.9.4...0.9.5)

## [0.9.4](https://github.com/Constantiner/cram-md5-digest-js/tree/0.9.4) (2018-10-09)

[Full Changelog](https://github.com/Constantiner/cram-md5-digest-js/compare/0.9.3...0.9.4)

## [0.9.3](https://github.com/Constantiner/cram-md5-digest-js/tree/0.9.3) (2018-10-05)

[Full Changelog](https://github.com/Constantiner/cram-md5-digest-js/compare/0.9.2...0.9.3)

## [0.9.2](https://github.com/Constantiner/cram-md5-digest-js/tree/0.9.2) (2018-06-07)

[Full Changelog](https://github.com/Constantiner/cram-md5-digest-js/compare/0.9.1...0.9.2)

## [0.9.1](https://github.com/Constantiner/cram-md5-digest-js/tree/0.9.1) (2018-05-30)

[Full Changelog](https://github.com/Constantiner/cram-md5-digest-js/compare/0.9.0...0.9.1)

## [0.9.0](https://github.com/Constantiner/cram-md5-digest-js/tree/0.9.0) (2018-05-16)

[Full Changelog](https://github.com/Constantiner/cram-md5-digest-js/compare/4b4ab48e52069f394e175d78403eef075af77cf8...0.9.0)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


src/cramMd5Digest.ts
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Lookup<T, K extends keyof any, Else = never> = K extends keyof T ? T[K] : Else;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Tail<T extends any[]> = ((...t: T) => void) extends (x: any, ...u: infer U) => void ? U : never;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Func1 = (arg: any) => any;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ArgType<F, Else = never> = F extends (arg: infer A) => any ? A : Else;
type AsChain<F extends [Func1, ...Func1[]], G extends Func1[] = Tail<F>> = {
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	[K in keyof F]: (arg: ArgType<F[K]>) => ArgType<Lookup<G, K, any>, any>;
};
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type LastIndexOf<T extends any[]> = ((...x: T) => void) extends (y: any, ...z: infer U) => void ? U["length"] : never;
type Block = [number, number, number, number, number, number[]];
interface State {
	byteCount?: number;
	state0?: number;
	state1?: number;
	state2?: number;
	state3?: number;
	block?: number[];
	padding?: number[];
	digest?: number[];
}

const range = (from: number, to: number): number[] =>
		Array.from(
			{
				length: to - from + 1
			},
			(_, index) => index + from
		),
	range64 = range(0, 63),
	T = range64.map(i => (Math.abs(Math.sin(i + 1)) * 0x100000000) | 0),
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	pipe = <F extends [(arg: any) => any, ...Array<(arg: any) => any>]>(
		...fns: F & AsChain<F>
	): // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
	((arg: ArgType<F[0]>) => ReturnType<F[LastIndexOf<F>]>) => fns.reduceRight((f, g) => value => f(g(value))),
	getEmptyArray = (length: number): number[] =>
		Array.from({
			length
		}),
	copyState = (source: State, overridesObject: Partial<State> = {}): State =>
		Object.assign(
			{},
			source,
			{
				block: [...source.block],
				padding: [...source.padding]
			},
			overridesObject
		),
	init = (initObject: State): State =>
		copyState(initObject, {
			byteCount: 0,
			state0: 0x67452301,
			state1: 0xefcdab89,
			state2: 0x98badcfe,
			state3: 0x10325476
		}),
	initialInit = (): State =>
		init({
			padding: getEmptyArray(64).map((_, index) => (index === 0 ? 0x80 : 0)),
			block: getEmptyArray(64)
		}),
	nBitsGenerator = function* (nBits: number): Generator<number, void, boolean> {
		while (nBits >= 6) {
			nBits = (Math.floor(nBits / 6) - 1) * 6;
			yield nBits;
		}
	},
	utfTextValue = (code: number) => (nBits: number): number => 0x80 + ((code >>> nBits) & 0x3f),
	nBitsMatches = [
		[0x00000800, 11],
		[0x00010000, 16],
		[0x00200000, 21],
		[0x04000000, 26],
		[0x80000000, 31],
		[Infinity, 0]
	],
	initialNBits = (code: number): number => nBitsMatches.find(([hexValue]) => code < hexValue)[1],
	initialUtfTextForCodeMoreThan0x80 = (code: number, nBits: number): number =>
		((0xfe << nBits % 6) & 0xff) | (code >>> (Math.floor(nBits / 6) * 6)),
	utfTextForCodeMoreThan0x80 = (code: number): number[] =>
		((nBits: number): number[] => [
			initialUtfTextForCodeMoreThan0x80(code, nBits),
			...[...nBitsGenerator(nBits)].map(utfTextValue(code))
		])(initialNBits(code)),
	toUTFArray = (stringToEncode: string): number[] =>
		[...stringToEncode]
			.map(x => x.charCodeAt(0))
			.reduce((utfText, code) => [...utfText, ...(code < 0x80 ? [code] : utfTextForCodeMoreThan0x80(code))], []),
	toUnsigned = (x: number): number => (x + 0x100000000) % 0x100000000,
	lShift = (x: number, s: number): number => (x << s) | (x >>> (32 - s)),
	// F, G, H and I are basic MD5 functions.
	F = (X: number, Y: number, Z: number): number => toUnsigned((X & Y) | (~X & Z)),
	G = (X: number, Y: number, Z: number): number => toUnsigned((X & Z) | (Y & ~Z)),
	H = (X: number, Y: number, Z: number): number => toUnsigned(X ^ Y ^ Z),
	I = (X: number, Y: number, Z: number): number => toUnsigned(Y ^ (X | ~Z)),
	// FF, GG, HH, and II transformations for rounds 1-4.
	XX = (func: (X: number, Y: number, Z: number) => number) => (xi: number, s: number) => ([
		a,
		b,
		c,
		d,
		i,
		x
	]: Block): Block => [d, toUnsigned(lShift(a + func(b, c, d) + x[xi] + T[i], s) + b), b, c, i + 1, x],
	FF = XX(F),
	GG = XX(G),
	HH = XX(H),
	II = XX(I),
	modifyStates = stateObject => ([a, b, c, d]: Block): State =>
		copyState(stateObject, {
			state0: toUnsigned(stateObject.state0 + a),
			state1: toUnsigned(stateObject.state1 + b),
			state2: toUnsigned(stateObject.state2 + c),
			state3: toUnsigned(stateObject.state3 + d)
		}),
	x = (stateObject: State): number[] =>
		getEmptyArray(16).map(
			(_, j) =>
				((stateObject.block[j * 4 + 3] * 256 + stateObject.block[j * 4 + 2]) * 256 +
					stateObject.block[j * 4 + 1]) *
					256 +
				stateObject.block[j * 4]
		),
	//
	// MD5 basic transformation. Transforms state based on block.
	//
	transformBlock = (stateObject: State): State =>
		pipe(
			FF(0, 7),
			FF(1, 12),
			FF(2, 17),
			FF(3, 22),
			FF(4, 7),
			FF(5, 12),
			FF(6, 17),
			FF(7, 22),
			FF(8, 7),
			FF(9, 12),
			FF(10, 17),
			FF(11, 22),
			FF(12, 7),
			FF(13, 12),
			FF(14, 17),
			FF(15, 22),
			GG(1, 5),
			GG(6, 9),
			GG(11, 14),
			GG(0, 20),
			GG(5, 5),
			GG(10, 9),
			GG(15, 14),
			GG(4, 20),
			GG(9, 5),
			GG(14, 9),
			GG(3, 14),
			GG(8, 20),
			GG(13, 5),
			GG(2, 9),
			GG(7, 14),
			GG(12, 20),
			HH(5, 4),
			HH(8, 11),
			HH(11, 16),
			HH(14, 23),
			HH(1, 4),
			HH(4, 11),
			HH(7, 16),
			HH(10, 23),
			HH(13, 4),
			HH(0, 11),
			HH(3, 16),
			HH(6, 23),
			HH(9, 4),
			HH(12, 11),
			HH(15, 16),
			HH(2, 23),
			II(0, 6),
			II(7, 10),
			II(14, 15),
			II(5, 21),
			II(12, 6),
			II(3, 10),
			II(10, 15),
			II(1, 21),
			II(8, 6),
			II(15, 10),
			II(6, 15),
			II(13, 21),
			II(4, 6),
			II(11, 10),
			II(2, 15),
			II(9, 21),
			modifyStates(stateObject)
		)([stateObject.state0, stateObject.state1, stateObject.state2, stateObject.state3, 0, x(stateObject)] as Block),
	update = (inputArray: number[], inputLength: number = inputArray.length) => (stateObject: State): State =>
		inputArray.slice(0, inputLength).reduce((stateObject_, item) => {
			stateObject_.block[stateObject_.byteCount % 64] = item;
			if (++stateObject_.byteCount % 64 === 0) {
				stateObject_ = transformBlock(stateObject_);
			}
			return stateObject_;
		}, copyState(stateObject)),
	set32Little = (value: number, index: number) => (array: number[]): number[] =>
		range(0, 3).reduce((array_, i) => ((array_[index + i] = (value >>> (i * 8)) & 0xff), array_), [...array]),
	getBits = (stateObject: State): number[] =>
		pipe(
			set32Little(stateObject.byteCount * 8, 0),
			set32Little(Math.floor((stateObject.byteCount * 8) / 0x100000000), 4)
		)(getEmptyArray(8)),
	alignIndex = (index: number): number =>
		([
			[56, (ind: number): number => 120 - ind],
			[-Infinity, (ind: number): number => 56 - ind]
		].find((range: [number, (ind: number) => number]) => index >= range[0])[1] as (ind: number) => number)(index),
	getIndex = (stateObject: State): number => alignIndex(stateObject.byteCount % 64),
	updateWithIndex = (stateObject: State): State => update(stateObject.padding, getIndex(stateObject))(stateObject),
	getFinalDigestObject = (stateObject: State): State => ({
		digest: pipe(
			set32Little(stateObject.state0, 0),
			set32Little(stateObject.state1, 4),
			set32Little(stateObject.state2, 8),
			set32Little(stateObject.state3, 12)
		)(getEmptyArray(16)),
		padding: [...stateObject.padding],
		block: [...stateObject.block]
	}),
	finalDigest = (stateObject: State): State =>
		pipe(updateWithIndex, update(getBits(stateObject)), getFinalDigestObject)(stateObject),
	hexByte = (x: number): string => (x < 16 ? "0" : "") + x.toString(16),
	toHexString = (byteArray: number[]): string =>
		byteArray.reduce((accumulator, x) => accumulator + hexByte(x), "").toLowerCase(),
	finalHexDigest = (stateObject: State): string => toHexString(finalDigest(stateObject).digest),
	expandArrayTo64Length = (original: number[]): number[] => [
		...original,
		...(Array.from({
			length: 64 - original.length
		}).fill(0) as number[])
	],
	updateWithDigest = (stateObject: State): State => update(stateObject.digest)(stateObject),
	performCram = (passwordAsUTFArray: number[], cramKey: string, stateObject: State): string => {
		const paddedKey: number[] = expandArrayTo64Length(
			passwordAsUTFArray.length > 64
				? ((stateObject = pipe(update(passwordAsUTFArray), finalDigest, init)(stateObject)), stateObject.digest)
				: passwordAsUTFArray
		).map(sym => sym ^ 0x36);

		// H(K XOR ipad, text) -> digest
		return pipe(
			update(paddedKey),
			update(toUTFArray(cramKey)),
			finalDigest,
			init,
			update(paddedKey.map(sym => sym ^ 0x36 ^ 0x5c)),
			updateWithDigest,
			finalHexDigest
		)(stateObject);
	},
	/**
	 * Generates MD5 hash from passwordString and cramKey
	 *
	 * @param {string} passwordString Is a secret (password) to hash.
	 * @param {string} cramKey Is a key (digest) to hash password with.
	 * @returns {string} Is resulting hash.
	 */
	cramMd5Digest = (passwordString: string, cramKey: string): string =>
		performCram(toUTFArray(passwordString), cramKey, initialInit()),
	encodeTable = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/",
	getPaddingSymbol = (padding: boolean): "=" | "" => (padding ? "=" : ""),
	postBase64Actions = (base64Sequence: string, encodeTable: string, md5Binary: number[], padding: boolean): string =>
		base64Sequence +
		encodeTable.charAt(md5Binary[30] >>> 2) +
		encodeTable.charAt(((md5Binary[30] << 4) & 0x30) | (md5Binary[30 + 1] >>> 4)) +
		encodeTable.charAt((md5Binary[31] << 2) & 0x3c) +
		getPaddingSymbol(padding),
	base64EncodeIndexGenerator = (): number[] =>
		Array.from(
			{
				length: 10
			},
			(_, i) => i * 3
		),
	getBase64Sequence = (md5Binary: number[]): string =>
		base64EncodeIndexGenerator().reduce(
			(result, index) =>
				result +
				encodeTable.charAt(md5Binary[index] >>> 2) +
				encodeTable.charAt(((md5Binary[index] << 4) & 0x30) | (md5Binary[index + 1] >>> 4)) +
				encodeTable.charAt(((md5Binary[index + 1] << 2) & 0x3c) | (md5Binary[index + 2] >>> 6)) +
				encodeTable.charAt(md5Binary[index + 2] & 0x3f),
			""
		),
	base64Encode = (md5Binary: number[], padding: boolean): string =>
		postBase64Actions(getBase64Sequence(md5Binary), encodeTable, md5Binary, padding),
	/**
	 * Generates MD5 hash from passwordString and cramKey and encodes it in Base64.
	 *
	 * @param {string} passwordString Is a password to hash.
	 * @param {string} cramKey Is a key to hash password with.
	 * @param {boolean} [padding=false] If true there is Base64 padding (=) added.
	 * @returns {string} Is resulting hash.
	 */
	cramMd5DigestBase64 = (passwordString: string, cramKey: string, padding = false): string =>
		base64Encode(
			[...cramMd5Digest(passwordString, cramKey)].map(char => char.charCodeAt(0)),
			padding
		);

export { cramMd5Digest, cramMd5DigestBase64 };
